test/MagicString.js

```diff
@@ -1860,5 +1860,30 @@ describe('MagicString', () => {
 				},
 			);
 		});
+
+		it('with offset instance',() => {
+			const s = new MagicString('hello world');
+			const so = s.withOffset(6);
+
+			assert.strictEqual(s instanceof MagicString, true);
+			assert.strictEqual(so instanceof MagicString, true);
+		});
+
+		it('with offset',() => {
+			let s = new MagicString('hello world').withOffset(6);
+
+			assert.strictEqual(s.slice(0, 5), 'world');
+			assert.strictEqual(s.appendLeft(0, ',').toString(), 'hello ,world');
+			assert.strictEqual(s.appendRight(0, ' ').toString(), 'hello , world');
+			assert.strictEqual(s.update(-1, 0, '').toString(), 'hello, world');
+			assert.strictEqual(s.overwrite(0, 1, 'w').toString(), 'hello,world');
+
+			s = s.withOffset(9);
+			assert.strictEqual(s.slice(), 'ld');
+
+			s.offset = 10;
```

Comment:
I am also not sure about this usage, it feels like we should either make it always immutable (always call `.withOffset` to create a new instance) or always mutable (that brings us back to the original issue).

Reply:
Yeah, after a second thought, I think the original design might be better. We shouldn't think too much about immutability here, as the MagicString itself is designed to be mutable. Let's check back to a single class with the mutable `offset` property then. Sorry for the back and forth

Reply:
No matter, I will add an offset filed in the afternoon.